diff --git a/drivers/gpu/drm/panel/panel-sitronix-st7701.c b/drivers/gpu/drm/panel/panel-sitronix-st7701.c
index 320a2a8fd..197dddef6 100644
--- a/drivers/gpu/drm/panel/panel-sitronix-st7701.c
+++ b/drivers/gpu/drm/panel/panel-sitronix-st7701.c
@@ -294,18 +294,18 @@ static const struct drm_panel_funcs st7701_funcs = {
 static const struct drm_display_mode ts8550b_mode = {
 	.clock		= 27500,
 
-	.hdisplay	= 480,
-	.hsync_start	= 480 + 38,
-	.hsync_end	= 480 + 38 + 12,
-	.htotal		= 480 + 38 + 12 + 12,
+	.hdisplay	= 640,
+	.hsync_start	= 640 + 16,
+	.hsync_end	= 640 + 16 + 8,
+	.htotal		= 640 + 16 + 8,
 
-	.vdisplay	= 854,
-	.vsync_start	= 854 + 18,
-	.vsync_end	= 854 + 18 + 8,
-	.vtotal		= 854 + 18 + 8 + 4,
+	.vdisplay	= 480,
+	.vsync_start	= 480 + 14,
+	.vsync_end	= 480 + 14 + 8,
+	.vtotal		= 480 + 14 + 8,
 
-	.width_mm	= 69,
-	.height_mm	= 139,
+	.width_mm	= 57,
+	.height_mm	= 43,
 
 	.type = DRM_MODE_TYPE_DRIVER | DRM_MODE_TYPE_PREFERRED,
 };
@@ -398,7 +398,7 @@ static int st7701_dsi_remove(struct mipi_dsi_device *dsi)
 }
 
 static const struct of_device_id st7701_of_match[] = {
-	{ .compatible = "techstar,ts8550b", .data = &ts8550b_desc },
+	{ .compatible = "fascontek,fs028vg047", .data = &ts8550b_desc },
 	{ }
 };
 MODULE_DEVICE_TABLE(of, st7701_of_match);
