diff --git a/sound/soc/sunxi/sun4i-i2s.c b/sound/soc/sunxi/sun4i-i2s.c
index e4a261f2e643..66a256bca7a8 100644
--- a/sound/soc/sunxi/sun4i-i2s.c
+++ b/sound/soc/sunxi/sun4i-i2s.c
@@ -391,6 +391,7 @@ static int sun4i_i2s_set_clk_rate(struct snd_soc_dai *dai,
 		}
 	}
 
+/*
 	int max_percent_diff = 5;
 	int max_diff = (clk_rate / (100 / max_percent_diff));
 	
@@ -398,6 +399,7 @@ static int sun4i_i2s_set_clk_rate(struct snd_soc_dai *dai,
 		dev_err(dai->dev, "MCLK diff too big!");
 		return -EINVAL;
 	}
+*/
 
 	if (mclk_rate == 0) {
 		dev_err(dai->dev, "Unable to find MCLK");
