diff --git a/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi b/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi
index 435a1e66aa6a..171c5caef381 100644
--- a/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi
+++ b/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi
@@ -23,7 +23,7 @@ dcxo: dcxo-clk {
 
 	de: display-engine {
 		compatible = "allwinner,sun20i-d1-display-engine";
-		allwinner,pipelines = <&mixer0>, <&mixer1>;
+		allwinner,pipelines = <&mixer0>;
 		status = "disabled";
 	};
 
@@ -683,28 +683,6 @@ mixer0_out_tcon_top_mixer0: endpoint {
 			};
 		};
 
-		mixer1: mixer@5200000 {
-			compatible = "allwinner,sun20i-d1-de2-mixer-1";
-			reg = <0x5200000 0x100000>;
-			clocks = <&display_clocks CLK_BUS_MIXER1>,
-				 <&display_clocks CLK_MIXER1>;
-			clock-names = "bus", "mod";
-			resets = <&display_clocks RST_MIXER1>;
-
-			ports {
-				#address-cells = <1>;
-				#size-cells = <0>;
-
-				mixer1_out: port@1 {
-					reg = <1>;
-
-					mixer1_out_tcon_top_mixer1: endpoint {
-						remote-endpoint = <&tcon_top_mixer1_in_mixer1>;
-					};
-				};
-			};
-		};
-
 		dsi: dsi@5450000 {
 			compatible = "allwinner,sun20i-d1-mipi-dsi",
 				     "allwinner,sun50i-a100-mipi-dsi";
@@ -777,33 +755,6 @@ tcon_top_mixer0_out_tcon_tv0: endpoint@2 {
 					};
 				};
 
-				tcon_top_mixer1_in: port@2 {
-					reg = <2>;
-					#address-cells = <1>;
-					#size-cells = <0>;
-
-					tcon_top_mixer1_in_mixer1: endpoint@1 {
-						reg = <1>;
-						remote-endpoint = <&mixer1_out_tcon_top_mixer1>;
-					};
-				};
-
-				tcon_top_mixer1_out: port@3 {
-					reg = <3>;
-					#address-cells = <1>;
-					#size-cells = <0>;
-
-					tcon_top_mixer1_out_tcon_lcd0: endpoint@0 {
-						reg = <0>;
-						remote-endpoint = <&tcon_lcd0_in_tcon_top_mixer1>;
-					};
-
-					tcon_top_mixer1_out_tcon_tv0: endpoint@2 {
-						reg = <2>;
-						remote-endpoint = <&tcon_tv0_in_tcon_top_mixer1>;
-					};
-				};
-
 				tcon_top_hdmi_in: port@4 {
 					reg = <4>;
 
@@ -844,11 +795,6 @@ tcon_lcd0_in_tcon_top_mixer0: endpoint@0 {
 						reg = <0>;
 						remote-endpoint = <&tcon_top_mixer0_out_tcon_lcd0>;
 					};
-
-					tcon_lcd0_in_tcon_top_mixer1: endpoint@1 {
-						reg = <1>;
-						remote-endpoint = <&tcon_top_mixer1_out_tcon_lcd0>;
-					};
 				};
 
 				tcon_lcd0_out: port@1 {
@@ -887,11 +833,6 @@ tcon_tv0_in_tcon_top_mixer0: endpoint@0 {
 						reg = <0>;
 						remote-endpoint = <&tcon_top_mixer0_out_tcon_tv0>;
 					};
-
-					tcon_tv0_in_tcon_top_mixer1: endpoint@1 {
-						reg = <1>;
-						remote-endpoint = <&tcon_top_mixer1_out_tcon_tv0>;
-					};
 				};
 
 				tcon_tv0_out: port@1 {
