diff --git a/drivers/gpu/drm/panel/panel-newvision-nv3052c.c b/drivers/gpu/drm/panel/panel-newvision-nv3052c.c
index cf078f0d3..c784b5718 100644
--- a/drivers/gpu/drm/panel/panel-newvision-nv3052c.c
+++ b/drivers/gpu/drm/panel/panel-newvision-nv3052c.c
@@ -230,8 +230,10 @@ static const struct nv3052c_reg nv3052c_panel_regs[] = {
 	{ 0xff, 0x52 },
 	{ 0xff, 0x00 },
 	{ 0x36, 0x0a },
+	{ 0x3a, 0x60 },
 };
 
+
 static inline struct nv3052c *to_nv3052c(struct drm_panel *panel)
 {
 	return container_of(panel, struct nv3052c, panel);
@@ -427,15 +429,15 @@ static void nv3052c_remove(struct spi_device *spi)
 
 static const struct drm_display_mode ltk035c5444t_modes[] = {
 	{ /* 60 Hz */
-		.clock = 24000,
+		.clock = 21000,
 		.hdisplay = 640,
-		.hsync_start = 640 + 96,
-		.hsync_end = 640 + 96 + 16,
-		.htotal = 640 + 96 + 16 + 48,
+		.hsync_start = 640 + 34,
+		.hsync_end = 640 + 96 + 4,
+		.htotal = 640 + 96 + 4 + 20,
 		.vdisplay = 480,
-		.vsync_start = 480 + 5,
-		.vsync_end = 480 + 5 + 2,
-		.vtotal = 480 + 5 + 2 + 13,
+		.vsync_start = 480 + 12,
+		.vsync_end = 480 + 12 + 4,
+		.vtotal = 480 + 12 + 4 + 6,
 		.flags = DRM_MODE_FLAG_NHSYNC | DRM_MODE_FLAG_NVSYNC,
 	},
 	{ /* 50 Hz */
@@ -455,8 +457,8 @@ static const struct drm_display_mode ltk035c5444t_modes[] = {
 static const struct nv3052c_panel_info ltk035c5444t_panel_info = {
 	.display_modes = ltk035c5444t_modes,
 	.num_modes = ARRAY_SIZE(ltk035c5444t_modes),
-	.width_mm = 77,
-	.height_mm = 64,
+	.width_mm = 153,
+	.height_mm = 90,
 	.bus_format = MEDIA_BUS_FMT_RGB888_1X24,
 	.bus_flags = DRM_BUS_FLAG_DE_HIGH | DRM_BUS_FLAG_PIXDATA_DRIVE_NEGEDGE,
 };
